File: frontend/src/app/classes/progress.ts
import { Example } from './example'
import { Task } from './task'

export class Progress {
  public static sum(acc: number, progress: Progress): number {
    return acc + (progress.completion || 0)
  }

  public constructor(
    public readonly task: Task<Example>,
    public readonly completion: number | undefined,
  ) { }
}

File: backend/src/main/java/pl/edu/ug/inf/lexica/domain/Progress.java
package pl.edu.ug.inf.lexica.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.*;
import org.hibernate.annotations.GenericGenerator;

import javax.persistence.*;
import java.util.UUID;

@Data
@RequiredArgsConstructor
@NoArgsConstructor
@Entity
public class Progress {
    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonIgnore
    private UUID id;

    @NonNull
    @ManyToOne
    @JsonIgnoreProperties(value = {"name", "example", "description", "type", "active", "examples"})
    private Task task;

    @JsonIgnore
    @EqualsAndHashCode.Exclude
    @ManyToOne
    private User user;

    @NonNull
    private int completion;
}
